daemon/stats/collector.go

```diff
@@ -111,44 +121,88 @@ func (s *Collector) Run() {
 		onlineCPUs, err := s.getNumberOnlineCPUs()
 		if err != nil {
 			log.G(context.TODO()).Errorf("collecting system online cpu count: %v", err)
+			// prevent CPU burning in case of something wrong at system level
+			time.Sleep(s.interval)
 			continue
 		}
-
-		for _, pair := range pairs {
-			stats, err := s.supervisor.GetContainerStats(pair.container)
-
-			switch err.(type) {
-			case nil:
-				// Sample system CPU usage close to container usage to avoid
-				// noise in metric calculations.
-				systemUsage, err := s.getSystemCPUUsage()
-				if err != nil {
-					log.G(context.TODO()).WithError(err).WithField("container_id", pair.container.ID).Errorf("collecting system cpu usage")
-					continue
+		s.Lock()
+		s.cache.onlineCPUs = onlineCPUs
+		s.Unlock()
+
+		g, ctx := errgroup.WithContext(context.Background())
+		pairCh := make(chan publishersPair)
+
+		g.Go(func() error {
+			defer close(pairCh)
+			for _, pair := range pairs {
+				select {
+				case pairCh <- pair:
+				case <-ctx.Done():
+					return ctx.Err()
 				}
-
-				// FIXME: move to containerd on Linux (not Windows)
-				stats.CPUStats.SystemUsage = systemUsage
-				stats.CPUStats.OnlineCPUs = onlineCPUs
-
-				pair.publisher.Publish(*stats)
-
-			case errdefs.ErrConflict, errdefs.ErrNotFound:
-				// publish empty stats containing only name and ID if not running or not found
-				pair.publisher.Publish(types.StatsJSON{
-					Name: pair.container.Name,
-					ID:   pair.container.ID,
-				})
-
-			default:
-				log.G(context.TODO()).Errorf("collecting stats for %s: %v", pair.container.ID, err)
-				pair.publisher.Publish(types.StatsJSON{
-					Name: pair.container.Name,
-					ID:   pair.container.ID,
-				})
 			}
+			return nil
+		})
+
+		numWorkers := int(onlineCPUs)
+		if numWorkers == 0 {
+			numWorkers = runtime.NumCPU()
+		}
+		if numWorkers > len(pairs) {
+			numWorkers = len(pairs)
 		}
 
+		for i := 0; i < numWorkers; i++ {
+			g.Go(func() error {
+				for pair := range pairCh {
+					stats, err := s.getContainerStats(pair.container)
+					if err != nil {
+						return err
+					}
+					pair.publisher.Publish(*stats)
+
+					select {
+					case <-ctx.Done():
+						return ctx.Err()
+					default:
+					}
+				}
+				return nil
+			})
+		}
+		g.Wait()
 		time.Sleep(s.interval)
 	}
 }
+
+func (s *Collector) getContainerStats(c *container.Container) (*types.StatsJSON, error) {
+	stats, err := s.supervisor.GetContainerStats(c)
+	switch err.(type) {
+	case nil:
+		// Sample system CPU usage close to container usage to avoid
+		// noise in metric calculations.
+		systemUsage, err := s.getSystemCPUUsage()
+		if err != nil {
+			log.G(context.TODO()).WithError(err).WithField("container_id", c.ID).Errorf("collecting system cpu usage")
+			return nil, err
+		}
+		// FIXME: move to containerd on Linux (not Windows)
+		stats.CPUStats.SystemUsage = systemUsage
+		s.RLock()
+		stats.CPUStats.OnlineCPUs = s.cache.onlineCPUs
```

Comment:
Right, it's a problem. I wanted to cache it because `onlineCPUs` does not change often and I'm not sure if the cost of many concurrent `sched_getaffinity` syscalls on large systems.